STAGE B - Quiz Questions 17 to 25

In [26]:
# Import necessary libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_absolute_error
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso

In [1]:
from google.colab import files
uploaded = files.upload()

Saving energydata_complete.csv to energydata_complete.csv


In [27]:
df = pd.read_csv('energydata_complete.csv')
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

In [28]:
df.isna().sum()

date           0
Appliances     0
lights         0
T1             0
RH_1           0
T2             0
RH_2           0
T3             0
RH_3           0
T4             0
RH_4           0
T5             0
RH_5           0
T6             0
RH_6           0
T7             0
RH_7           0
T8             0
RH_8           0
T9             0
RH_9           0
T_out          0
Press_mm_hg    0
RH_out         0
Windspeed      0
Visibility     0
Tdewpoint      0
rv1            0
rv2            0
dtype: int64


## Question 17

### **From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the Root Mean Squared error in three D.P?**

In [30]:
# Extract features (T2) and target (T6)
X = df[['T2']]
y = df['T6']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create a linear regression model
model = LinearRegression()

# Fit the model to the training data
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the RMSE rounded to three decimal places
print(f"Root Mean Squared Error (RMSE): {rmse:.3f}")

Root Mean Squared Error (RMSE): 3.633


## Question 18

### **Remove the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Normalize the dataset using the MinMaxScaler (Hint: Use the MinMaxScaler fit_transform and transform methods on the train and test set respectively). Run a multiple linear regression using the training set. Answer the following questions:**

### What is the Mean Absolute Error (in three decimal places) for the  training set?

In [31]:
# Remove columns "date" and "lights"
df.drop(columns=['date', 'lights'], inplace=True)

# Separate features (X) and target (y)
X = df.drop(columns=['Appliances'])
y = df['Appliances']

# Split data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Normalize the dataset using MinMaxScaler
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Create and fit a multiple linear regression model
model = LinearRegression()
model.fit(X_train_scaled, y_train)

# Make predictions on the training set
y_pred_train = model.predict(X_train_scaled)

# Calculate Mean Absolute Error (MAE) for the training set
mae_train = mean_absolute_error(y_train, y_pred_train)

# Print the MAE rounded to three decimal places
print(f"Mean Absolute Error (MAE) for the training set: {mae_train:.3f}")


Mean Absolute Error (MAE) for the training set: 53.742


### Question 19

### What is the Root Mean Squared Error (in three decimal places) for the training set?

In [32]:
# Calculate Root Mean Squared Error (RMSE) for the training set
rmse_train = np.sqrt(mean_squared_error(y_train, y_pred_train))

print(f"Root Mean Squared Error (RMSE) for the training set: {rmse_train:.3f}")

Root Mean Squared Error (RMSE) for the training set: 95.216


## Question 20

### What is the Mean Absolute Error (in three decimal places) for test set?

In [33]:
# Make predictions on the test set
y_pred_test = model.predict(X_test_scaled)

# Calculate Mean Absolute Error (MAE) for the test set
mae_test = mean_absolute_error(y_test, y_pred_test)

print(f"Mean Absolute Error (MAE) for the test set: {mae_test:.3f}")

Mean Absolute Error (MAE) for the test set: 53.643


## Question 21

### What is the Root Mean Squared Error (in three decimal places) for test set?

In [34]:
# Calculate Root Mean Squared Error (RMSE) for the test set
rmse_test = np.sqrt(mean_squared_error(y_test, y_pred_test))

print(f"Root Mean Squared Error (RMSE) for the test set: {rmse_test:.3f}")

Root Mean Squared Error (RMSE) for the test set: 93.640


### Question 22

### Did the Model above overfit to the training set

In [35]:
# Determine if the model overfit the training set
overfit = "Yes" if rmse_train < rmse_test else "No"

print(f"Did the model overfit to the training set? {overfit}")

Did the model overfit to the training set? No


### Question 23

### Train a ridge regression model with default parameters. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [36]:
# Create and fit a Ridge regression model
model_ridge = Ridge()  # Default alpha = 1.0
model_ridge.fit(X_train, y_train)

# Make predictions on the test set
y_pred_test = model_ridge.predict(X_test)

# Calculate RMSE for the test set
rmse_test = mean_squared_error(y_test, y_pred_test, squared=False)

# Print the RMSE rounded to three decimal places
print(f"Root Mean Squared Error (RMSE) for the test set: {rmse_test:.3f}")

Root Mean Squared Error (RMSE) for the test set: 93.641


Answer : Yes

 There is a change to the root mean squared error (RMSE) when evaluated on the test set for the ridge regression model as compared to the Linear Regression model

 Root Mean Squared Error (RMSE) for the test set: 93.640 (Linear Regression)

 Root Mean Squared Error (RMSE) for the test set: 93.641 (Ridge Regression)

### Question 24

### Train a lasso regression model with default value and obtain the new feature weights with it. How many of the features have non-zero feature weights?

In [37]:
# Create and fit a Lasso regression model
model = Lasso()  # Default alpha = 1.0
model.fit(X_train, y_train)

# Get the feature weights (coefficients)
feature_weights = model.coef_

# Count the number of non-zero feature weights
non_zero_features = sum(feature_weights != 0)

print(f"Number of features with non-zero weights: {non_zero_features}")


Number of features with non-zero weights: 21


### Question 25

### What is the new RMSE with the Lasso Regression on the test set?

In [38]:
# Create and fit a Lasso regression model
model = Lasso()  # Default alpha = 1.0
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred_test = model.predict(X_test)

# Calculate RMSE for the test set
rmse_test = mean_squared_error(y_test, y_pred_test, squared=False)

# Print the RMSE rounded to three decimal places
print(f"Root Mean Squared Error (RMSE) for the test set: {rmse_test:.3f}")


Root Mean Squared Error (RMSE) for the test set: 93.892
